subprojects/plugins/src/main/java/org/gradle/api/tasks/Upload.java

```diff
@@ -19,140 +19,141 @@
 import org.gradle.api.Action;
+import org.gradle.api.InvalidUserCodeException;
 import org.gradle.api.artifacts.Configuration;
-import org.gradle.api.artifacts.PublishException;
 import org.gradle.api.artifacts.dsl.RepositoryHandler;
 import org.gradle.api.file.FileCollection;
 import org.gradle.api.internal.ConventionTask;
-import org.gradle.api.internal.artifacts.ArtifactPublicationServices;
-import org.gradle.api.internal.artifacts.ArtifactPublisher;
-import org.gradle.api.internal.artifacts.Module;
-import org.gradle.api.internal.artifacts.configurations.ConfigurationInternal;
-import org.gradle.api.internal.artifacts.repositories.PublicationAwareRepository;
-import org.gradle.internal.Transformers;
-import org.gradle.internal.deprecation.DeprecationLogger;
-import org.gradle.util.internal.ConfigureUtil;
+import org.gradle.api.internal.DocumentationRegistry;
 import org.gradle.work.DisableCachingByDefault;
 
 import javax.annotation.Nullable;
 import javax.inject.Inject;
 import java.io.File;
-import java.util.List;
-
-import static org.gradle.util.internal.CollectionUtils.collect;
 
 /**
- * Uploads the artifacts of a {@link Configuration} to a set of repositories.
+ * This task is <strong>no longer supported</strong> and will <strong>throw an exception</strong> if you try to use it.
+ * It is preserved solely for backwards compatibility and may be removed in a future version.
  *
- * @deprecated This class is scheduled for removal in Gradle 8.0. To upload artifacts, use the maven-publish plugin instead.
+ * @deprecated This class is scheduled for removal in a future version. To upload artifacts, use the maven-publish plugin instead.
  */
-@Deprecated
+@Deprecated // TODO:Finalize Upload Removal - Issue #21439
 @DisableCachingByDefault(because = "Produces no cacheable output")
-public class Upload extends ConventionTask {
-
-    private Configuration configuration;
-    private boolean uploadDescriptor;
-    private File descriptorDestination;
-    private RepositoryHandler repositories;
-
+public abstract class Upload extends ConventionTask {
+    /**
+     * Do not use this method, it is for internal use only.
+     * @deprecated This class is scheduled for removal in a future version, this method <strong>should not be used</strong>.
+     */
+    @Deprecated
     @Inject
-    protected ArtifactPublicationServices getPublicationServices() {
-        throw new UnsupportedOperationException();
-    }
+    protected abstract DocumentationRegistry getDocumentationRegistry();
 
     @TaskAction
     protected void upload() {
-        getLogger().info("Publishing configuration: {}", configuration);
-        DeprecationLogger.deprecateTaskType(Upload.class, getPath()).willBeRemovedInGradle8().withUpgradeGuideSection(7, "upload_task_deprecation").nagUser();
-
-        Module module = ((ConfigurationInternal) configuration).getModule();
-
-        ArtifactPublisher artifactPublisher = getPublicationServices().createArtifactPublisher();
```

Comment:
Seems like some of the uploading logic is still left around here. Specifically, `ArtifactPublisher` and its implementations, as well as many of the files in `org.gradle.internal.component.external.ivypublish`.

Reply:
`ArtifactPublisher` and the types in that package are available as a build service - I don't know what other external clients could be using it, and I think it would need to go through a deprecation cycle to be removed, even though it is `internal`.

Reply:
@big-guy - What do you think?  Should we start a new PR to deprecate all these?

Reply:
Let's get this PR merged first, but we don't need to go through a deprecation cycle with `ArtifactPublisher`.  None of these are public APIs or a build service.